week3/homework/src/utilities.js

```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+const dataDealer = require('./data/data_dealer');
+const read = dataDealer.read;
+const write = dataDealer.write;
+const shortid = require('shortid');
+const TodoItem = require('./todo_obj_constructor');
+
+const validation = function(req) {
+  const reqBodyLength = Object.keys(req.body).length;
+  const reqToDo = req.body.todo || {};
+  let reqDescription = reqToDo.description || {};
+
+  if (typeof reqDescription === 'string') {
+    reqDescription = reqDescription.trim();
+  }
+
+  if (
+    reqBodyLength !== 1 ||
+        reqDescription === {} ||
+        typeof reqDescription !== 'string' ||
+        reqDescription.length === 0
+  ) {
+    return (reqDescription = '');
+  }
+  return reqDescription;
+};
+
+const findToDo = function(CurrentList, reqId) {
+  let toDo = CurrentList.find(todo => todo.id === reqId);
+  if (toDo === undefined) {
+    throw new Error(`The To-Do item with ID:${reqId} is not existed`);
+  }
+  return toDo;
+};
+
+const createToDo = function(req, res) {
+  const toDoDescription = validation(req);
+  if (toDoDescription === '') {
+    res.status(202).json({ Notification: 'The posted To-Do is not valid' });
+    return;
+  }
+  const todoObject = new TodoItem(shortid.generate(), toDoDescription, false);
+  read('./data/todolist.json', 'utf8')
+    .then(result => {
+      const CurrentList = JSON.parse(result);
+      CurrentList.push(todoObject);
+      const updatedList = JSON.stringify(CurrentList, null, 2);
+      write('./data/todolist.json', updatedList);
+      res.status(201).json({ Notification: 'new To-DO is added' });
+    })
+    .catch(err => res.status(500).json({ Error: err.message }));
```

Comment:
the `read` method errors mostly are going to be about a the file we have, which means nothing for the user

I would rather hard code the error to be something as `Oops something went wrong!` but generally speaking your errors here sometimes are about existence of the todo so you could pass the `response` object to send the response to the client from the `findTodo` function instead of throwing.

Or as a solution you could pass the `next` method and let it pass to the error handler route with it's correct status code.

Reply:
2 good ideas . 
So in general it is good to decrease throwing errors while another options are there.